mobile/src/main/java/org/openhab/habdroid/background/BackgroundTasksManager.java

```diff
@@ -66,6 +66,10 @@ public void onReceive(Context context, Intent intent) {
 
     public static void scheduleWorker(Context context, String key) {
         SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
+        if (Constants.PREFERENCE_DEMOMODE.equals(key)) {
```

Comment:
That wasn't exactly what I meant. I meant to check `prefs.getBoolean(Constants.PREFERENCE_DEMOMODE, false)` here and call `cancelAllWorkByTag(key)` here if it's true, and additionally...

Reply:
The preference is set to `false` after `setOnPreferenceChangeListener()` returned `true`. So calling `scheduleWorker()` and fetching the pref there returns the old value.

Reply:
No, because `scheduleWorker()` in the prefs fragment schedules the call via a Handler for precisely that reason ;-)

Reply:
Done, but the issue still exists.

Reply:
But how? I'd expect this to happen:
- demo mode is enabled
- work is canceled through `scheduleWorker()`
- respective work items have their state change from FAILED to CANCELED
- change is picked up by `NotificationUpdateObserver`
- no item in FAILED state anymore -> notification is removed

What am I missing there?

Reply:
Answer to self: the FAILED -> CANCELED transition doesn't happen, because FAILED is a terminal state (only ENQUEUED, BLOCKED and RUNNING can transition to FAILED)

Fixed by pruning the worker DB in that case: https://github.com/maniac103/openhab.android/commit/8c05ad256db0febc0744ae0e2a4b1ceee680969b

Reply:
Thanks. Should be fine to merge now.